File: src/main/java/imp/typechecker/TypeChecker.java
// TypeChecker.java
package imp.typechecker;

import imp.ast.*;
import imp.ast.condition.*;
import imp.ast.expression.*;
import imp.ast.expression.array.NewArrayExpression;
import imp.ast.expression.binary.bool.compare.*;
import imp.ast.expression.binary.bool.logic.AndExpression;
import imp.ast.expression.binary.bool.logic.ImpliesExpression;
import imp.ast.expression.binary.bool.logic.OrExpression;
import imp.ast.expression.binary.integer.*;
import imp.ast.expression.bool.ExistsExpression;
import imp.ast.expression.bool.ForallExpression;
import imp.ast.expression.constant.bool.FalseExpression;
import imp.ast.expression.constant.bool.TrueExpression;
import imp.ast.expression.constant.integer.IntExpression;
import imp.ast.expression.unary.bool.NotExpression;
import imp.ast.expression.unary.integer.NegExpression;
import imp.ast.method.*;
import imp.ast.statement.*;
import imp.ast.typing.*;
import imp.ast.typing.data.DataType;
import imp.ast.typing.data.array.ArrayType;
import imp.ast.typing.data.array.BoolArray;
import imp.ast.typing.data.array.IntArray;
import imp.ast.typing.data.value.BoolType;
import imp.ast.typing.data.value.IntType;
import imp.ast.expression.Identifier;
import imp.ast.ASTVisitor;

import java.util.*;

public class TypeChecker {

    public static void checkTypes(Program program) {
        TypeCheckVisitor visitor = new TypeCheckVisitor();
        program.accept(visitor);
    }

    private static class TypeCheckVisitor extends ASTVisitor {
        private final TypeSymbolTable symTab;
        private final Map<Identifier, FunctionType> functionSymTab;
        private Type lastType;

        public TypeCheckVisitor() {
            symTab = new TypeSymbolTable();
            functionSymTab = new HashMap<>();
            lastType = null;
        }

        @Override
        public void visit(Program program) {
            // First, collect all method signatures
            for (MethodDeclaration method : program.methods()) {
                collectMethodSignature(method);
            }
            // Now, type-check each method
            for (MethodDeclaration method : program.methods()) {
                method.accept(this);
            }
        }

        private void collectMethodSignature(MethodDeclaration methodDeclaration) {
            List<DataType> paramTypes = methodDeclaration.parameterList().parameters().stream()
                    .map(Parameter::type)
                    .toList();

            // Determine return type
            DataType returnType = Optional.ofNullable(methodDeclaration.returnValue()).map(ReturnValue::type).orElse(VoidType.getInstance());
            // Create FunctionType
            FunctionType methodType = new FunctionType(paramTypes, returnType);
            // Add to function symbol table
            functionSymTab.put(methodDeclaration.name(), methodType);
        }

        @Override
        public void visit(MethodDeclaration methodDeclaration) {
            symTab.pushState();

            // Add parameters to symbol table
            methodDeclaration.parameterList().accept(this);

            // **Add return variable to symbol table if present**
            if (methodDeclaration.returnValue() != null) {
                methodDeclaration.returnValue().accept(this);
            }

            // Process conditions (requires and ensures)
            methodDeclaration.conditionList().accept(this);

            // Visit method body
            methodDeclaration.methodBody().accept(this);

            symTab.popState();
        }

        @Override
        public void visit(ParameterList parameterList) {
            for (Parameter param : parameterList.parameters()) {
                param.accept(this);
            }
        }

        @Override
        public void visit(Parameter parameter) {
            if (symTab.containsKey(parameter.name())) {
                throw new TypeError("Parameter " + parameter.name() + " already declared.");
            }
            symTab.put(parameter.name(), parameter.type());
        }

        @Override
        public void visit(ReturnValue returnValue) {
            Identifier returnVarName = returnValue.name();
            Type returnType = returnValue.type();
            // Check for redeclaration
            if (symTab.containsKey(returnVarName)) {
                throw new TypeError("Return variable " + returnVarName + " already declared as a parameter.");
            }
            symTab.put(returnVarName, returnType);
        }

        @Override
        public void visit(ConditionList conditionList) {
            for (ConditionClause clause : conditionList.conditions()) {
                clause.accept(this);
            }
        }

        @Override
        public void visit(EnsuresClause ensuresClause) {
            ensuresClause.expr().accept(this);
            if (!(lastType instanceof BoolType)) {
                throw new TypeError("Ensures clause must be of type bool.");
            }
        }

        @Override
        public void visit(RequiresClause requiresClause) {
            requiresClause.expr().accept(this);
            if (!(lastType instanceof BoolType)) {
                throw new TypeError("Requires clause must be of type bool.");
            }
        }

        @Override
        public void visit(InvariantList invariantList) {
            for (Invariant invariant : invariantList.invariants()) {
                invariant.accept(this);
            }
        }

        @Override
        public void visit(Invariant invariant) {
            invariant.expression().accept(this);
            if (!(lastType instanceof BoolType)) {
                throw new TypeError("Invariant expression must be of type bool.");
            }
        }

        @Override
        public void visit(MethodBody methodBody) {
            for (Statement stmt : methodBody.statements()) {
                stmt.accept(this);
            }
        }

        private Type visitAndGet(ASTNode node) {
            node.accept(this);
            return lastType;
        }

        @Override
        public void visit(VariableDeclaration variableDeclaration) {
            Identifier varName = variableDeclaration.variableName();
            if (symTab.containsKey(varName)) {
                throw new TypeError("Variable " + varName + " already declared.");
            }
            Type declaredType = visitAndGet(variableDeclaration.declaredType());

            symTab.put(varName, declaredType);

            if (variableDeclaration.initializer().isPresent()) {
                Type initializerType = visitAndGet(variableDeclaration.initializer().get());

                if (initializerType != declaredType) {
                    throw new TypeError("Cannot assign " + initializerType + " to variable " + varName + " of type " + declaredType);
                }
            }
        }

        @Override
        public void visit(AssignStatement assignStatement) {
            assignStatement.lhs().accept(this);
            Type lhsType = lastType;

            assignStatement.expression().accept(this);
            Type rhsType = lastType;

            if (rhsType != lhsType) {
                throw new TypeError("Cannot assign " + rhsType + " to " + assignStatement.lhs());
            }
        }

        @Override
        public void visit(IfStatement ifStatement) {
            ifStatement.condition().accept(this);
            if (!(lastType instanceof BoolType)) {
                throw new TypeError("If condition must be of type bool.");
            }

            symTab.pushState();
            ifStatement.thenBlock().accept(this);
            symTab.popState();

            if (ifStatement.elseBlock().isPresent()) {
                symTab.pushState();
                ifStatement.elseBlock().get().accept(this);
                symTab.popState();
            }
        }

        @Override
        public void visit(WhileStatement whileStatement) {
            whileStatement.condition().accept(this);
            if (!(lastType instanceof BoolType)) {
                throw new TypeError("While condition must be of type bool.");
            }

            if (!whileStatement.invariants().invariants().isEmpty()) {
                whileStatement.invariants().accept(this);
            }

            symTab.pushState();
            whileStatement.body().accept(this);
            symTab.popState();
        }

        @Override
        public void visit(BlockStatement blockStatement) {
            symTab.pushState();
            for (Statement stmt : blockStatement.statements()) {
                stmt.accept(this);
            }
            symTab.popState();
        }

        @Override
        public void visit(VarRefExpression varRefExpression) {
            Identifier varName = varRefExpression.variableName();
            Type varType = symTab.get(varName);
            if (varType == null) {
                throw new TypeError("Variable " + varName + " is not declared.");
            }
            lastType = varType;
            if (!(lastType instanceof DataType)) {
                throw new TypeError("Variable " + varName + " is not a data type.");
            }
            varRefExpression.setType((DataType) lastType);
        }

        @Override
        public void visit(ArrayRefExpression arrayRefExpression) {
            Identifier arrayName = arrayRefExpression.arrayName();
            Type arrayType = symTab.get(arrayName);
            if (arrayType == null) {
                throw new TypeError("Array " + arrayName + " is not declared.");
            }
            if (!(arrayType instanceof ArrayType)) {
                throw new TypeError(arrayName + " is not an array.");
            }

            arrayRefExpression.indexExpr().accept(this);
            if (!(lastType instanceof IntType)) {
                throw new TypeError("Array index must be of type int.");
            }

            if (arrayType instanceof BoolArray) {
                lastType = BoolType.getInstance();
                arrayRefExpression.setElementType(BoolType.getInstance());
            } else {
                arrayRefExpression.setElementType(IntType.getInstance());
                lastType = IntType.getInstance();
            }
        }

        public void visit(BinaryOpExpression binaryOpExpression) {
            binaryOpExpression.left().accept(this);
            Type leftType = lastType;

            binaryOpExpression.right().accept(this);
            Type rightType = lastType;

            String op = binaryOpExpression.operatorSymbol();
            Type resultType = getBinaryOpResultType(op, leftType, rightType);

            if (resultType == null) {
                throw new TypeError("Invalid operands for operator '" + op + "' in expression '" + binaryOpExpression + "': " + leftType + ", " + rightType);
            }
            lastType = resultType;
        }

        public void visit(UnaryExpression unaryExpression) {
            unaryExpression.expression().accept(this);
            Type operandType = lastType;

            String op = unaryExpression.operatorSymbol();
            Type resultType = getUnaryOpResultType(op, operandType);

            if (resultType == null) {
                throw new TypeError("Invalid operand for operator '" + op + "' in expression '" + unaryExpression + "': " + operandType);
            }
            lastType = resultType;
        }

        @Override
        public void visit(IntExpression intExpression) {
            lastType = IntType.getInstance();
        }

        @Override
        public void visit(FalseExpression falseExpression) {
            lastType = BoolType.getInstance();
        }

        @Override
        public void visit(TrueExpression trueExpression) {
            lastType = BoolType.getInstance();
        }

        @Override
        public void visit(NotExpression notExpression) {
            visit((UnaryExpression) notExpression);
        }

        @Override
        public void visit(NegExpression negExpression) {
            visit((UnaryExpression) negExpression);
        }

        @Override
        public void visit(ImpliesExpression impliesExpression) {
            visit((BinaryOpExpression) impliesExpression);
        }

        @Override
        public void visit(OrExpression orExpression) {
            visit((BinaryOpExpression) orExpression);
        }

        @Override
        public void visit(MulExpression mulExpression) {
            visit((BinaryOpExpression) mulExpression);
        }

        @Override
        public void visit(EqExpression eqExpression) {
            visit((BinaryOpExpression) eqExpression);
        }

        @Override
        public void visit(LessThanOrEqualExpression lessThanOrEqualExpression) {
            visit((BinaryOpExpression) lessThanOrEqualExpression);
        }

        @Override
        public void visit(SubExpression subExpression) {
            visit((BinaryOpExpression) subExpression);
        }

        @Override
        public void visit(ModExpression modExpression) {
            visit((BinaryOpExpression) modExpression);
        }

        @Override
        public void visit(GreaterThanOrEqualExpression greaterThanOrEqualExpression) {
            visit((BinaryOpExpression) greaterThanOrEqualExpression);
        }

        @Override
        public void visit(AddExpression addExpression) {
            visit((BinaryOpExpression) addExpression);
        }

        @Override
        public void visit(LessThanExpression lessThanExpression) {
            visit((BinaryOpExpression) lessThanExpression);
        }

        @Override
        public void visit(DivExpression divExpression) {
            visit((BinaryOpExpression) divExpression);
        }

        @Override
        public void visit(GreaterThanExpression greaterThanExpression) {
            visit((BinaryOpExpression) greaterThanExpression);
        }

        @Override
        public void visit(AndExpression andExpression) {
            visit((BinaryOpExpression) andExpression);
        }

        @Override
        public void visit(Condition condition) {
            condition.expression().accept(this);
            if (!(lastType instanceof BoolType)) {
                throw new TypeError("Condition expression must be of type bool.");
            }
        }

        @Override
        public void visit(BoolType boolType) {
            lastType = BoolType.getInstance();
        }

        @Override
        public void visit(IntArray intArray) {
            lastType = IntArray.getInstance();
        }

        @Override
        public void visit(BoolArray boolArray) {
            lastType = BoolArray.getInstance();
        }

        @Override
        public void visit(VoidType voidType) {
            lastType = VoidType.getInstance();
        }

        @Override
        public void visit(IntType intType) {
            lastType = IntType.getInstance();
        }

        @Override
        public void visit(FuncCallStatement funcCallStatement) {
            funcCallStatement.funcCallExpression().accept(this);
        }

        @Override
        public void visit(FuncCallExpression funcCallExpression) {
            Identifier funcName = funcCallExpression.identifier();
            FunctionType funcType = functionSymTab.get(funcName);
            if (funcType == null) {
                throw new TypeError("Unknown function: " + funcName);
            }

            List<Expression> args = funcCallExpression.arguments();
            if (args.size() != funcType.getParameterTypes().size()) {
                throw new TypeError("Function " + funcName + " expects " + funcType.getParameterTypes().size() + " arguments, but got " + args.size());
            }

            for (int i = 0; i < args.size(); i++) {
                Type argType = visitAndGet(args.get(i));
                Type expectedType = funcType.getParameterTypes().get(i);
                if (argType != expectedType) {
                    throw new TypeError("Function " + funcName + " argument " + (i + 1) + " expects type " + expectedType + ", but got " + argType);
                }
            }
            funcCallExpression.setType(funcType);

            lastType = funcType.getReturnType();
        }

        @Override
        public void visit(NewArrayExpression newArrayExpression) {
            newArrayExpression.sizeExpr().accept(this);
            if (!(lastType instanceof IntType)) {
                throw new TypeError("Array size must be of type int.");
            }
            if (newArrayExpression.elementType() instanceof BoolType) {
                lastType = BoolArray.getInstance();
            } else {
                lastType = IntArray.getInstance();
            }
        }

        @Override
        public void visit(ArrayLengthExpression arrayLengthExpression) {
            arrayLengthExpression.arrayExpr().accept(this);
            if (!(lastType instanceof ArrayType)) {
                throw new TypeError("Cannot get length of non-array type.");
            }
            lastType = IntType.getInstance();
        }

        @Override
        public void visit(ForallExpression forallExpression) {
            symTab.pushState();
            symTab.put(forallExpression.variable(), forallExpression.getType());
            forallExpression.body().accept(this);
            if (!(lastType instanceof BoolType)) {
                throw new TypeError("Body of forall expression must be of type bool.");
            }
            symTab.popState();
            lastType = BoolType.getInstance();
        }

        @Override
        public void visit(ExistsExpression existsExpression) {
            symTab.pushState();
            symTab.put(existsExpression.variable(), existsExpression.getType());
            existsExpression.body().accept(this);
            if (!(lastType instanceof BoolType)) {
                throw new TypeError("Body of exists expression must be of type bool.");
            }
            symTab.popState();
            lastType = BoolType.getInstance();
        }

        private Type getBinaryOpResultType(String op, Type leftType, Type rightType) {
            if (op.matches("\\+|\\-|\\*|/|%")) {
                if (leftType instanceof IntType && rightType instanceof IntType) {
                    return IntType.getInstance();
                }
            } else if (op.matches("&&|\\|\\|")) {
                if (leftType instanceof BoolType && rightType instanceof BoolType) {
                    return BoolType.getInstance();
                }
            } else if (op.matches("<|>|<=|>=")) {
                if (leftType instanceof IntType && rightType instanceof IntType) {
                    return BoolType.getInstance();
                }
            } else if (op.matches("==")) {
                if (leftType instanceof IntType && rightType instanceof IntType) {
                    return BoolType.getInstance();
                }
                if (leftType instanceof BoolType && rightType instanceof BoolType) {
                    return BoolType.getInstance();
                }
            }
            else if (op.equals("==>")) { // Implies operator
                if (leftType instanceof BoolType && rightType instanceof BoolType) {
                    return BoolType.getInstance();
                }
            }
            return null;
        }

        private Type getUnaryOpResultType(String op, Type operandType) {
            if (op.equals("-") && operandType instanceof IntType) {
                return IntType.getInstance();
            } else if (op.equals("!") && operandType instanceof BoolType) {
                return BoolType.getInstance();
            }
            return null;
        }
    }

    // Custom exception for type errors
    public static class TypeError extends RuntimeException {
        public TypeError(String message) {
            super("TypeError: " + message);
        }
    }
}

File: src/type_checker.cpp
#include "ivy/type_checker.hpp"

#include <optional>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

#include "ivy/errors.hpp"
#include "ivy/pretty_printer.hpp"

namespace ivy {
namespace {

class TypeSymbolTable final {
  public:
    TypeSymbolTable() {
        stack_.emplace_back();
    }

    void PushState() {
        stack_.push_back(stack_.back());
    }

    void PopState() {
        if (stack_.size() <= 1U) {
            throw std::logic_error("TypeSymTab: cannot pop last symbol table");
        }
        stack_.pop_back();
    }

    [[nodiscard]] bool Contains(const std::string &key) const {
        return stack_.back().find(key) != stack_.back().end();
    }

    void Put(const std::string &key, const DataType value) {
        stack_.back()[key] = value;
    }

    [[nodiscard]] std::optional<DataType> Get(const std::string &key) const {
        const auto it = stack_.back().find(key);
        if (it == stack_.back().end()) {
            return std::nullopt;
        }
        return it->second;
    }

  private:
    std::vector<std::unordered_map<std::string, DataType>> stack_;
};

class TypeCheckerImpl final {
  public:
    void Check(Program &program) {
        CollectMethodSignatures(program);
        for (MethodDeclaration &method : program.methods) {
            CheckMethod(method);
        }
    }

  private:
    void CollectMethodSignatures(const Program &program) {
        for (const MethodDeclaration &method : program.methods) {
            std::vector<DataType> parameter_types;
            parameter_types.reserve(method.parameters.size());
            for (const Parameter &parameter : method.parameters) {
                parameter_types.push_back(parameter.type);
            }
            const DataType return_type =
                method.return_value.has_value() ? method.return_value->type : DataType::kVoid;
            function_symbol_table_[method.name] = FunctionType{std::move(parameter_types), return_type};
        }
    }

    void CheckMethod(MethodDeclaration &method) {
        symbol_table_.PushState();

        for (const Parameter &parameter : method.parameters) {
            if (symbol_table_.Contains(parameter.name)) {
                throw TypeError("Parameter " + parameter.name + " already declared.");
            }
            symbol_table_.Put(parameter.name, parameter.type);
        }

        if (method.return_value.has_value()) {
            const Parameter &return_value = *method.return_value;
            if (symbol_table_.Contains(return_value.name)) {
                throw TypeError("Return variable " + return_value.name + " already declared as a parameter.");
            }
            symbol_table_.Put(return_value.name, return_value.type);
        }

        for (ConditionClause &clause : method.conditions) {
            const DataType condition_type = EvalExpression(*clause.expression);
            if (condition_type != DataType::kBool) {
                if (clause.kind == ConditionClause::Kind::kEnsures) {
                    throw TypeError("Ensures clause must be of type bool.");
                }
                throw TypeError("Requires clause must be of type bool.");
            }
        }

        for (StatementPtr &statement : method.body) {
            CheckStatement(*statement);
        }

        symbol_table_.PopState();
    }

    void CheckStatement(Statement &statement) {
        std::visit(
            [&](auto &node) {
                using T = std::decay_t<decltype(node)>;
                if constexpr (std::is_same_v<T, Statement::Assign>) {
                    const DataType lhs_type = EvalExpression(*node.lhs);
                    const DataType rhs_type = EvalExpression(*node.rhs);
                    if (lhs_type != rhs_type) {
                        throw TypeError("Cannot assign " + ToString(rhs_type) + " to " +
                                        PrettyPrint(*node.lhs));
                    }
                } else if constexpr (std::is_same_v<T, Statement::If>) {
                    const DataType condition_type = EvalExpression(*node.condition);
                    if (condition_type != DataType::kBool) {
                        throw TypeError("If condition must be of type bool.");
                    }

                    symbol_table_.PushState();
                    for (StatementPtr &then_statement : node.then_block) {
                        CheckStatement(*then_statement);
                    }
                    symbol_table_.PopState();

                    if (node.else_block.has_value()) {
                        symbol_table_.PushState();
                        for (StatementPtr &else_statement : *node.else_block) {
                            CheckStatement(*else_statement);
                        }
                        symbol_table_.PopState();
                    }
                } else if constexpr (std::is_same_v<T, Statement::While>) {
                    const DataType condition_type = EvalExpression(*node.condition);
                    if (condition_type != DataType::kBool) {
                        throw TypeError("While condition must be of type bool.");
                    }

                    for (ExpressionPtr &invariant : node.invariants) {
                        const DataType invariant_type = EvalExpression(*invariant);
                        if (invariant_type != DataType::kBool) {
                            throw TypeError("Invariant expression must be of type bool.");
                        }
                    }

                    symbol_table_.PushState();
                    for (StatementPtr &body_statement : node.body) {
                        CheckStatement(*body_statement);
                    }
                    symbol_table_.PopState();
                } else if constexpr (std::is_same_v<T, Statement::Block>) {
                    symbol_table_.PushState();
                    for (StatementPtr &block_statement : node.statements) {
                        CheckStatement(*block_statement);
                    }
                    symbol_table_.PopState();
                } else if constexpr (std::is_same_v<T, Statement::VarDecl>) {
                    if (symbol_table_.Contains(node.variable)) {
                        throw TypeError("Variable " + node.variable + " already declared.");
                    }

                    symbol_table_.Put(node.variable, node.declared_type);

                    if (node.initializer.has_value()) {
                        const DataType initializer_type = EvalExpression(**node.initializer);
                        if (initializer_type != node.declared_type) {
                            throw TypeError("Cannot assign " + ToString(initializer_type) + " to variable " +
                                            node.variable + " of type " + ToString(node.declared_type));
                        }
                    }
                } else if constexpr (std::is_same_v<T, Statement::FuncCallStmt>) {
                    static_cast<void>(EvalExpression(*node.call));
                }
            },
            statement.node);
    }

    [[nodiscard]] DataType EvalExpression(Expression &expression) {
        return std::visit(
            [&](auto &node) -> DataType {
                using T = std::decay_t<decltype(node)>;
                if constexpr (std::is_same_v<T, Expression::VarRef>) {
                    const std::optional<DataType> variable_type = symbol_table_.Get(node.name);
                    if (!variable_type.has_value()) {
                        throw TypeError("Variable " + node.name + " is not declared.");
                    }
                    node.resolved_type = variable_type;
                    return *variable_type;
                } else if constexpr (std::is_same_v<T, Expression::ArrayRef>) {
                    const std::optional<DataType> array_type = symbol_table_.Get(node.name);
                    if (!array_type.has_value()) {
                        throw TypeError("Array " + node.name + " is not declared.");
                    }
                    if (!IsArrayType(*array_type)) {
                        throw TypeError(node.name + " is not an array.");
                    }

                    const DataType index_type = EvalExpression(*node.index);
                    if (index_type != DataType::kInt) {
                        throw TypeError("Array index must be of type int.");
                    }

                    node.element_type =
                        *array_type == DataType::kBoolArray ? DataType::kBool : DataType::kInt;
                    return *node.element_type;
                } else if constexpr (std::is_same_v<T, Expression::FuncCall>) {
                    const auto function_it = function_symbol_table_.find(node.name);
                    if (function_it == function_symbol_table_.end()) {
                        throw TypeError("Unknown function: " + node.name);
                    }
                    const FunctionType &function_type = function_it->second;
                    if (node.arguments.size() != function_type.parameter_types.size()) {
                        throw TypeError("Function " + node.name + " expects " +
                                        std::to_string(function_type.parameter_types.size()) +
                                        " arguments, but got " + std::to_string(node.arguments.size()));
                    }

                    for (std::size_t i = 0; i < node.arguments.size(); ++i) {
                        const DataType argument_type = EvalExpression(*node.arguments[i]);
                        const DataType expected_type = function_type.parameter_types[i];
                        if (argument_type != expected_type) {
                            throw TypeError("Function " + node.name + " argument " + std::to_string(i + 1U) +
                                            " expects type " + ToString(expected_type) + ", but got " +
                                            ToString(argument_type));
                        }
                    }

                    node.resolved_type = function_type;
                    return function_type.return_type;
                } else if constexpr (std::is_same_v<T, Expression::Unary>) {
                    const DataType operand_type = EvalExpression(*node.operand);
                    const std::optional<DataType> result_type = UnaryResultType(node.op, operand_type);
                    if (!result_type.has_value()) {
                        throw TypeError("Invalid operand for operator '" + ToString(node.op) +
                                        "' in expression '" + PrettyPrint(expression) +
                                        "': " + ToString(operand_type));
                    }
                    return *result_type;
                } else if constexpr (std::is_same_v<T, Expression::Binary>) {
                    const DataType left_type = EvalExpression(*node.left);
                    const DataType right_type = EvalExpression(*node.right);
                    const std::optional<DataType> result_type =
                        BinaryResultType(node.op, left_type, right_type);
                    if (!result_type.has_value()) {
                        throw TypeError("Invalid operands for operator '" + ToString(node.op) +
                                        "' in expression '" + PrettyPrint(expression) +
                                        "': " + ToString(left_type) + ", " + ToString(right_type));
                    }
                    return *result_type;
                } else if constexpr (std::is_same_v<T, Expression::IntLiteral>) {
                    return DataType::kInt;
                } else if constexpr (std::is_same_v<T, Expression::BoolLiteral>) {
                    return DataType::kBool;
                } else if constexpr (std::is_same_v<T, Expression::Quantified>) {
                    symbol_table_.PushState();
                    symbol_table_.Put(node.variable, node.type);
                    const DataType body_type = EvalExpression(*node.body);
                    symbol_table_.PopState();

                    if (body_type != DataType::kBool) {
                        if (node.forall) {
                            throw TypeError("Body of forall expression must be of type bool.");
                        }
                        throw TypeError("Body of exists expression must be of type bool.");
                    }
                    return DataType::kBool;
                } else if constexpr (std::is_same_v<T, Expression::NewArray>) {
                    const DataType size_type = EvalExpression(*node.size);
                    if (size_type != DataType::kInt) {
                        throw TypeError("Array size must be of type int.");
                    }
                    return node.element_type == DataType::kBool ? DataType::kBoolArray : DataType::kIntArray;
                } else if constexpr (std::is_same_v<T, Expression::ArrayLength>) {
                    const DataType array_type = EvalExpression(*node.array);
                    if (!IsArrayType(array_type)) {
                        throw TypeError("Cannot get length of non-array type.");
                    }
                    return DataType::kInt;
                }
                throw TypeError("Unknown expression node");
            },
            expression.node);
    }

    [[nodiscard]] static std::optional<DataType> UnaryResultType(const UnaryOp op,
                                                                 const DataType operand_type) {
        if (op == UnaryOp::kNeg && operand_type == DataType::kInt) {
            return DataType::kInt;
        }
        if (op == UnaryOp::kNot && operand_type == DataType::kBool) {
            return DataType::kBool;
        }
        return std::nullopt;
    }

    [[nodiscard]] static std::optional<DataType> BinaryResultType(const BinaryOp op, const DataType left_type,
                                                                  const DataType right_type) {
        switch (op) {
        case BinaryOp::kAdd:
        case BinaryOp::kSub:
        case BinaryOp::kMul:
        case BinaryOp::kDiv:
        case BinaryOp::kMod:
            if (left_type == DataType::kInt && right_type == DataType::kInt) {
                return DataType::kInt;
            }
            return std::nullopt;
        case BinaryOp::kAnd:
        case BinaryOp::kOr:
        case BinaryOp::kImplies:
            if (left_type == DataType::kBool && right_type == DataType::kBool) {
                return DataType::kBool;
            }
            return std::nullopt;
        case BinaryOp::kLess:
        case BinaryOp::kLessOrEqual:
        case BinaryOp::kGreater:
        case BinaryOp::kGreaterOrEqual:
            if (left_type == DataType::kInt && right_type == DataType::kInt) {
                return DataType::kBool;
            }
            return std::nullopt;
        case BinaryOp::kEqual:
            if ((left_type == DataType::kInt && right_type == DataType::kInt) ||
                (left_type == DataType::kBool && right_type == DataType::kBool)) {
                return DataType::kBool;
            }
            return std::nullopt;
        }
        return std::nullopt;
    }

    TypeSymbolTable symbol_table_;
    std::unordered_map<std::string, FunctionType> function_symbol_table_;
};

} // namespace

void CheckTypes(Program &program) {
    TypeCheckerImpl checker;
    checker.Check(program);
}

} // namespace ivy
